framework/libra-framework/sources/modified_source/stake.move

```diff
@@ -1365,21 +1379,40 @@ module diem_framework::stake {
 
         let current_vals = get_current_validators();
 
-        if (vector::length(&proposed) <= min) {
 
-            proposed = get_sorted_vals_by_props(vector::length(&current_vals) / 2);
+        let is_performant_below_f_4 = vector::length(&performant) <= ( 2 * (min_f+1) + 1);
 
-        };
+        let is_proposed_below_f_3 = vector::length(&proposed) <= ( 2 * min_f + 1);
 
+        // hail mary.
+        // there may be something wrong with performance metrics or evaluation
+        // do nothing
+        if (is_proposed_below_f_3 && is_performant_below_f_4) {
+          return current_vals
+        };
 
-        // It's not clear that there could be another failure, but fully backstop it by having the same validtor set.
-        if (vector::length(&proposed) <= min) {
+        // happy case, not near failure
+        if (!is_proposed_below_f_3 && !is_performant_below_f_4) return proposed;
+        // the proposed validators per the algo is to low.
+        // and the performant validators from previous epoch are at a healthy
+        // number, just failover to the performant validators
+        if (is_proposed_below_f_3 && !is_performant_below_f_4) {
+          return performant
+        };
 
-          return current_vals
+        // The proposed number of validators is not near failure
+        // but the previous epoch's performing nodes is below healthy
+        // pick whichever one has the most number
+        if (!is_proposed_below_f_3 && is_performant_below_f_4) {
+          if (vector::length(&performant) > vector::length(&proposed)) {
+            return performant
+          };
+          return proposed
```

Comment:
@0o-de-lally 
Why not take a greedy approach, cover all `performant` and then pick top ranked order `proposed`?

Reply:
It's a good idea. These failover states are always attack vectors, so need to consider carefully. Why don't you ticket this please for future consideration? Then you can also tag yourself as the feature owner 🦾